256 create table

uptimeDBA opened the pull request:
A bit of a moving target this one, and I broke something using git but I think I fixed it.
Some links are dependent on the constraints documentation that is still under review.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/296)
<!-- Reviewable:end -->

jseldess commented on create-table.md:
```diff
@@ -1,22 +1,111 @@
 ---
 title: CREATE TABLE
 toc: false
+toc_not_nested: true
```
I know the contributing.md file asked for `toc_not_nested` at some point, but it's no longer needed.

jseldess commented on create-table.md:
```diff
@@ -1,22 +1,111 @@
-|-----------|-------------|
-|  |  |
\ No newline at end of file
+## Semantics
```
Because the sql diagram parameters do not always have specific, helpful names, I've been avoiding this type of list in favor of a "Usage" section. But the parameter names are sufficiently clear in this case, so I understand why you've built in this section. 

For consistency, though, please rename it to "Parameters" and try to do this as a table.

jseldess commented on create-table.md:
```diff
@@ -1,22 +1,111 @@
+An optional comma separated list of any [table level constraints](data-definition.html#table-level-constraints). Constraint names have their own namespace within a table's definition so are required to be unique for the table but may have the same name as columns or indexes, however this is not recommended. 
+
+
+## Usage
```
Since we're defining the parameters above, I don't think we need a "Usage" section. I'd rather have more detailed examples (see next comment). The notes about primary key and replication zones are helpful, so perhaps just add them to the general description at the top?

jseldess commented on create-table.md:
```diff
@@ -1,22 +1,111 @@
+
+By default, tables are created in the default replication zone but can be placed into a specific replication zone. See the discussion on [Creating a Replication Zone for a Table](configure-replication-zones.html#create-a-replication-zone-for-a-table) for more information.
+
+## Examples
```
Instead of one example containing all complexities, can you try to break this into a few separate examples?

jseldess commented on create-table.md:
```diff
@@ -1,22 +1,111 @@
+- [`RENAME TABLE`](rename-table.html)
+- [`TRUNCATE`](truncate.html)
+- [`SHOW TABLES`](show-tables.html)
+- [`SHOW CREATE`]()
```
Broken link here. Though I don't think we have a page for `SHOW CREATE TABLE` yet, so you should probably just remove this one.

uptimeDBA replied:
I missed the second `SHOW CREATE` broken link.

jseldess commented:
Thanks for this work, @uptimeDBA! I added some comments and will find the right devs to review for you. As with the other PR, once this is good-to-go from a correctness standpoint, I'll merge and do a copy edit in another PR.

knz commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[create-table.md, line 31 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHzZGOhr3aEElAo4AWU:-KHzZGOhr3aEElAo4AWV:-1875536046) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L31)):*
> ```Markdown
> The [name](data-definition.html#identifiers) (optionally schema-qualified) of the table to be created. Table names are required to be unique within a database.
> 
> The `UPSERT` and `INSERT INTO ... ON CONFLICT` statements make use of a table called `EXCLUDED` to handle data conflicts during execution. It's therefore not recommended to use the name `EXCLUDED` for any of your tables.
> ```

This is a relatively obscure feature for most readers. If you want to make this point, please support it with examples (and/or counter-examples).

---

*[create-table.md, line 87 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHzZxc0Ayohdkn_Cpva:-KHzZxc0Ayohdkn_Cpvb:1959668593) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L87)):*
> ```Markdown
> 
> 
> As the syntax diagram suggests, you can create a table without any columns like you can in PostgreSQL. The table does have a column called `rowid` that was added because the table does not have a `PRIMARY KEY` or `UNIQUE` defined.
> ```

You have already refered above to the fact that a rowid column is implicitly created if there is no primary key. This is clear. However it is not clear here why the UNIQUE constraint makes a difference. Please clarify more.

Also it would be interesting to make a comment as to why/how a table with no columns could be useful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/296)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[create-table.md, line 31 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHzZGOhr3aEElAo4AWU:-KI3OT-vxPAGtfy4WF-m:947895886) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L31)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This is a relatively obscure feature for most readers. If you want to make this point, please support it with examples (and/or counter-examples).
</details>
If you do have a table named "excluded", can you work around the conflict with `insert into excluded as excluded_table on conflict foo do update set excluded_table.x=excluded.x`? As long as that works it might be better to document this workaround in the `insert on conflict` docs instead of a much more prominent warning against tables named "excluded"

---

*[create-table.md, line 87 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHzZxc0Ayohdkn_Cpva:-KI3QtgAxkqHgk-166jR:1408746697) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L87)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> You have already refered above to the fact that a rowid column is implicitly created if there is no primary key. This is clear. However it is not clear here why the UNIQUE constraint makes a difference. Please clarify more.
> 
> Also it would be interesting to make a comment as to why/how a table with no columns could be useful.
</details>
The rowid column is created whenever there is no primary key; it doesn't matter whether there are unique columns.

```
root@:26257> create table t2 (a int unique);
CREATE TABLE
root@:26257> show columns from t2;
+-------+------+-------+----------------+
| Field | Type | Null  |    Default     |
+-------+------+-------+----------------+
| a     | INT  | true  | NULL           |
| rowid | INT  | false | unique_rowid() |
+-------+------+-------+----------------+

```

A table with no columns is useful for finding bugs in client libraries (i just found one in the go lib/pq driver), but i'm not sure if it's ever legitimately useful (it's also possible to create such a table by dropping the last column). 

```
root@:26257> create table t (a int);
CREATE TABLE
root@:26257> insert into t values (1);
INSERT 1
root@:26257> alter table t drop column a;
ALTER TABLE
root@:26257> select * from t;
pq: unexpected DataRow in simple query execution
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/296)*
<!-- Sent from Reviewable.io -->

uptimeDBA commented:
Review status: 0 of 33 files reviewed at latest revision, 7 unresolved discussions.

---

*[create-table.md, line 4 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHwRVRlv5KbdcxkTeVy-r9-4:-KI5rJkq1XKmywr5rz-H:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L4)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> I know the contributing.md file asked for `toc_not_nested` at some point, but it's no longer needed.
</details>
Done.

---

*[create-table.md, line 20 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHwRVRlv5KbdcxkTeVy-r9-20:-KI5uWESFcqwVz4d6aTw:1161300322) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L20)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Because the sql diagram parameters do not always have specific, helpful names, I've been avoiding this type of list in favor of a "Usage" section. But the parameter names are sufficiently clear in this case, so I understand why you've built in this section. 
> 
> 
> 
> For consistency, though, please rename it to "Parameters" and try to do this as a table.  
</details>
I've converted the Semantics section into a table and renamed it Parameters. Looks ok in html but it's very messy in markdown as the formatting options seem limited inside a table. Take a look and see what you think.

---

*[create-table.md, line 31 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHzZGOhr3aEElAo4AWU:-KI6QieCqGFbnGi5lzn2:-882940281) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L31)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> If you do have a table named "excluded", can you work around the conflict with `insert into excluded as excluded_table on conflict foo do update set excluded_table.x=excluded.x`? As long as that works it might be better to document this workaround in the `insert on conflict` docs instead of a much more prominent warning against tables named "excluded"
</details>
Good work around. Should probably be mentioned as part of the `INSERT ... ON CONFLICT DO UPDATE` documentation.

---

*[create-table.md, line 49 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHwRVRlv5KbdcxkTeVy-r9-49:-KI6QF8WKaLTy0AXXUu1:-405925879) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L49)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Since we're defining the parameters above, I don't think we need a "Usage" section. I'd rather have more detailed examples (see next comment). The notes about primary key and replication zones are helpful, so perhaps just add them to the general description at the top?
</details>
"Usage" section removed with examples added to the Examples section. Replication Zone note added to the top description.

---

*[create-table.md, line 87 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHzZxc0Ayohdkn_Cpva:-KI6HpfcoOeJH9hpRKad:-1771399855) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L87)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The rowid column is created whenever there is no primary key; it doesn't matter whether there are unique columns.
> 
> ```
> root@:26257> create table t2 (a int unique);
> CREATE TABLE
> root@:26257> show columns from t2;
> +-------+------+-------+----------------+
> | Field | Type | Null  |    Default     |
> +-------+------+-------+----------------+
> | a     | INT  | true  | NULL           |
> | rowid | INT  | false | unique_rowid() |
> +-------+------+-------+----------------+
> 
> ```
> 
> A table with no columns is useful for finding bugs in client libraries (i just found one in the go lib/pq driver), but i'm not sure if it's ever legitimately useful (it's also possible to create such a table by dropping the last column). 
> 
> ```
> root@:26257> create table t (a int);
> CREATE TABLE
> root@:26257> insert into t values (1);
> INSERT 1
> root@:26257> alter table t drop column a;
> ALTER TABLE
> root@:26257> select * from t;
> pq: unexpected DataRow in simple query execution
> ```
</details>
I've removed the no-column table example.  I was intrigued that you could even do this. In 20+ years as an operational DBA/architect, I've never seen a no-column table. As you say, it's so left-field, it's not worth a mention.

---

*[create-table.md, line 109 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/docs/296#-KHwRVRlv5KbdcxkTeVy-r9-109:-KI6Plgc_D6DsDx29Gl3:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/725852e5b16e8ad636c8030e39aa078bc986237c/create-table.md#L109)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Broken link here. Though I don't think we have a page for `SHOW CREATE TABLE` yet, so you should probably just remove this one.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/296)*
<!-- Sent from Reviewable.io -->

jseldess commented:
LGTM, Paul. I'll merge now so we can include in this week's release notes. I'll open a separate PR to tweak some language/formatting, but nothing urgent.
